File: web/src/lib/goalMath.ts
import { endOfMonth, isBefore, startOfDay } from "date-fns";

export type GoalInputMode = "daily" | "weekly" | "monthly";
export type PaceMode = "calendar_days" | "weekdays";

export type GoalDraft = {
  dailyGoalHours: number;
  weeklyGoalHours: number;
  monthlyGoalHours: number;
  paceMode: PaceMode;
  daysPerWeek: number;
  effectiveFrom: Date;
};

function clampWhole(value: number, min: number, max: number): number {
  return Math.min(Math.max(Math.round(value), min), max);
}

function normalizeHours(value: number): number {
  if (!Number.isFinite(value) || value < 0) {
    return 0;
  }
  return Number((Math.round(value * 3600) / 3600).toFixed(2));
}

function hoursToSeconds(value: number): number {
  return Math.max(0, Math.round(value * 3600));
}

function secondsToHours(value: number): number {
  return Number((value / 3600).toFixed(2));
}

export function countActiveDaysInMonth(monthAnchor: Date, paceMode: PaceMode): number {
  const monthStart = new Date(monthAnchor.getFullYear(), monthAnchor.getMonth(), 1);
  const monthEnd = endOfMonth(monthStart);

  let cursor = monthStart;
  let count = 0;
  while (!isBefore(monthEnd, cursor)) {
    const isWeekday = cursor.getDay() !== 0 && cursor.getDay() !== 6;
    if (paceMode === "calendar_days" || isWeekday) {
      count += 1;
    }
    cursor = new Date(cursor.getFullYear(), cursor.getMonth(), cursor.getDate() + 1);
  }

  return Math.max(count, 1);
}

export function deriveGoalDraft(params: {
  inputMode: GoalInputMode;
  inputGoalHours: number;
  paceMode: PaceMode;
  daysPerWeek: number;
  effectiveFrom: Date;
}): GoalDraft {
  const daysPerWeek = clampWhole(params.daysPerWeek, 1, 7);
  const activeDays = countActiveDaysInMonth(params.effectiveFrom, params.paceMode);
  const inputGoalSeconds = hoursToSeconds(params.inputGoalHours);

  let dailyGoalSeconds = 0;
  let weeklyGoalSeconds = 0;
  let monthlyGoalSeconds = 0;

  if (params.inputMode === "monthly") {
    monthlyGoalSeconds = inputGoalSeconds
    dailyGoalSeconds = Math.round(monthlyGoalSeconds / activeDays)
    weeklyGoalSeconds = Math.round(dailyGoalSeconds * daysPerWeek)
  } else if (params.inputMode === "weekly") {
    weeklyGoalSeconds = inputGoalSeconds
    dailyGoalSeconds = Math.round(weeklyGoalSeconds / daysPerWeek)
    monthlyGoalSeconds = Math.round(dailyGoalSeconds * activeDays)
  } else {
    dailyGoalSeconds = inputGoalSeconds
    weeklyGoalSeconds = Math.round(dailyGoalSeconds * daysPerWeek)
    monthlyGoalSeconds = Math.round(dailyGoalSeconds * activeDays)
  }

  return {
    dailyGoalHours: normalizeHours(secondsToHours(dailyGoalSeconds)),
    weeklyGoalHours: normalizeHours(secondsToHours(weeklyGoalSeconds)),
    monthlyGoalHours: normalizeHours(secondsToHours(monthlyGoalSeconds)),
    paceMode: params.paceMode,
    daysPerWeek,
    effectiveFrom: params.effectiveFrom
  };
}

export function countRemainingTargetDays(fromDate: Date, paceMode: PaceMode): number {
  const today = startOfDay(fromDate);
  const monthEnd = endOfMonth(today);

  let cursor = today;
  let count = 0;
  while (!isBefore(monthEnd, cursor)) {
    const isWeekday = cursor.getDay() !== 0 && cursor.getDay() !== 6;
    if (paceMode === "calendar_days" || isWeekday) {
      count += 1;
    }
    cursor = new Date(cursor.getFullYear(), cursor.getMonth(), cursor.getDate() + 1);
  }

  return Math.max(count, 1);
}

export function buildRecommendedPace(params: {
  monthlyGoalHours: number;
  monthlyHoursSoFar: number;
  paceMode: PaceMode;
  daysPerWeek: number;
  fromDate: Date;
}) {
  const remainingHours = Math.max(params.monthlyGoalHours - params.monthlyHoursSoFar, 0);
  const targetDays = countRemainingTargetDays(params.fromDate, params.paceMode);
  const daysPerWeek = clampWhole(params.daysPerWeek, 1, 7);
  const recDaily = remainingHours / targetDays;
  const recWeekly = recDaily * daysPerWeek;

  return {
    targetDays,
    remainingHours: normalizeHours(remainingHours),
    recDaily: normalizeHours(recDaily),
    recWeekly: normalizeHours(recWeekly)
  };
}

export function formatGoalInput(value: number): string {
  const normalized = normalizeHours(value);
  return String(Number(normalized.toFixed(2)));
}

File: ios/CampusTracker/Features/Settings/GoalMath.swift
import Foundation

struct GoalDraft {
    let dailyGoalHours: Double
    let weeklyGoalHours: Double
    let monthlyGoalHours: Double
    let paceMode: GoalPaceMode
    let daysPerWeek: Int
}

struct RecommendedPace {
    let remainingHours: Double
    let targetDays: Int
    let dailyHours: Double
    let weeklyHours: Double
}

enum GoalMath {
    static func deriveDraft(
        inputMode: GoalInputMode,
        inputGoalHours: Double,
        paceMode: GoalPaceMode,
        daysPerWeek: Int,
        effectiveFrom: Date
    ) -> GoalDraft {
        let normalizedDaysPerWeek = min(max(daysPerWeek, 1), 7)
        let activeDays = activeDaysInMonth(for: effectiveFrom, paceMode: paceMode)
        let inputGoalSeconds = max(Int(round(inputGoalHours * 3600)), 0)

        let dailyGoalSeconds: Int
        let weeklyGoalSeconds: Int
        let monthlyGoalSeconds: Int

        switch inputMode {
        case .monthly:
            monthlyGoalSeconds = inputGoalSeconds
            dailyGoalSeconds = Int(round(Double(monthlyGoalSeconds) / Double(activeDays)))
            weeklyGoalSeconds = Int(round(Double(dailyGoalSeconds * normalizedDaysPerWeek)))
        case .weekly:
            weeklyGoalSeconds = inputGoalSeconds
            dailyGoalSeconds = Int(round(Double(weeklyGoalSeconds) / Double(normalizedDaysPerWeek)))
            monthlyGoalSeconds = Int(round(Double(dailyGoalSeconds * activeDays)))
        case .daily:
            dailyGoalSeconds = inputGoalSeconds
            weeklyGoalSeconds = Int(round(Double(dailyGoalSeconds * normalizedDaysPerWeek)))
            monthlyGoalSeconds = Int(round(Double(dailyGoalSeconds * activeDays)))
        }

        return GoalDraft(
            dailyGoalHours: normalizeHours(Double(dailyGoalSeconds) / 3600),
            weeklyGoalHours: normalizeHours(Double(weeklyGoalSeconds) / 3600),
            monthlyGoalHours: normalizeHours(Double(monthlyGoalSeconds) / 3600),
            paceMode: paceMode,
            daysPerWeek: normalizedDaysPerWeek
        )
    }

    static func buildRecommendedPace(
        monthlyGoalHours: Double,
        monthlyHoursSoFar: Double,
        paceMode: GoalPaceMode,
        daysPerWeek: Int,
        from now: Date
    ) -> RecommendedPace {
        let targetDays = remainingTargetDays(from: now, paceMode: paceMode)
        let remainingHours = max(monthlyGoalHours - monthlyHoursSoFar, 0)
        let normalizedDaysPerWeek = min(max(daysPerWeek, 1), 7)
        let dailyHours = remainingHours / Double(targetDays)

        return RecommendedPace(
            remainingHours: normalizeHours(remainingHours),
            targetDays: targetDays,
            dailyHours: normalizeHours(dailyHours),
            weeklyHours: normalizeHours(dailyHours * Double(normalizedDaysPerWeek))
        )
    }

    static func activeDaysInMonth(for monthAnchor: Date, paceMode: GoalPaceMode) -> Int {
        let calendar = Calendar.current
        let monthStart = calendar.date(from: calendar.dateComponents([.year, .month], from: monthAnchor)) ?? monthAnchor
        let daysInMonth = calendar.range(of: .day, in: .month, for: monthStart)?.count ?? 30

        if paceMode == .calendarDays {
            return max(daysInMonth, 1)
        }

        var count = 0
        for offset in 0..<daysInMonth {
            guard let day = calendar.date(byAdding: .day, value: offset, to: monthStart) else { continue }
            if !calendar.isDateInWeekend(day) {
                count += 1
            }
        }

        return max(count, 1)
    }

    static func normalizeHours(_ value: Double) -> Double {
        guard value.isFinite, value >= 0 else { return 0 }
        return (value * 100).rounded() / 100
    }

    private static func remainingTargetDays(from now: Date, paceMode: GoalPaceMode) -> Int {
        let calendar = Calendar.current
        let start = calendar.startOfDay(for: now)
        let monthStart = calendar.date(from: calendar.dateComponents([.year, .month], from: start)) ?? start
        let monthEnd = calendar.date(byAdding: DateComponents(month: 1, day: -1), to: monthStart) ?? start

        var count = 0
        var cursor = start
        while cursor <= monthEnd {
            if paceMode == .calendarDays || !calendar.isDateInWeekend(cursor) {
                count += 1
            }
            guard let next = calendar.date(byAdding: .day, value: 1, to: cursor), next > cursor else {
                break
            }
            cursor = next
        }

        return max(count, 1)
    }
}
